presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java

```diff
@@ -15,40 +15,66 @@
 
 import com.facebook.presto.spi.type.NamedTypeSignature;
 import com.facebook.presto.spi.type.ParameterKind;
-import com.facebook.presto.spi.type.StandardTypes;
 import com.facebook.presto.spi.type.Type;
 import com.facebook.presto.spi.type.TypeManager;
 import com.facebook.presto.spi.type.TypeSignature;
 import com.facebook.presto.spi.type.TypeSignatureParameter;
-import com.facebook.presto.type.UnknownType;
+import com.google.common.base.VerifyException;
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
 
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
+import java.util.Set;
 
 import static com.facebook.presto.type.TypeCalculation.calculateLiteralValue;
+import static com.facebook.presto.type.TypeRegistry.isCovariantTypeBase;
+import static com.facebook.presto.type.UnknownType.UNKNOWN;
+import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.base.Preconditions.checkState;
-import static com.google.common.collect.Iterables.getLast;
+import static com.google.common.base.Verify.verify;
+import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
 import static java.util.stream.Collectors.toList;
 import static java.util.stream.Collectors.toMap;
 
+/**
+ * Determines whether, and how, a callsite matches a generic function signature.
+ * Which is equivalent to finding assignments for the variables in the generic signature,
+ * such that all of the function's declared parameters are super types of the corresponding
+ * arguments, and also satisfy the declared constraints (such as a given type parameter must
+ * bind to an orderable type)
+ * <p>
+ * This implementation has made assumptions. When any of the assumptions is not satisfied, it will fail loudly.
+ * <p><ul>
+ * <li>A type cannot have both type parameter and literal parameter.
+ * <li>A literal parameter cannot be be used across types. see {@link #checkNoLiteralVariableUsageAcrossTypes(TypeSignature, Map)}.
+ * </ul><p>
+ */
 public class SignatureBinder
 {
+    private static final int SOLVE_ITERATION_LIMIT = 4;
```

Comment:
With the iterative solver, you need 2 for anything with generic in it. For expressions involving lambda (which is under progress), you will need 3.

The number here is chosen rather arbitrarily. The purpose of this constant is to avoid infinite loop. I'll add comments.